### PR TITLE
Refactor to reduce coupling of base UtBot implementation with Spring-specific features

### LINKED ISSUE #2358: Avoid `when`s by `applicationContext` by introducing `UTBotComponentFactory`
**Description**

Right now codebase contains a lot of code frgaments like this:
```kotlin
when (applicationContext) {
    is SpringApplicationContext -> doSomethingSpringSpecific()
    else -> doSomethingDefault()
}
```

**Action plan**

Utilize abstract factory design pattern by making `ApplicationContext` an abstract factory responsible for creating different components of UTBot (possible renaming it to `UTBotComponentFactory`).

Note: it's not as simple as it may look, since many interfaces need to be extracted and moved to `utbot-framework-api` and constructor and method parameters of both `SimpleUtBotComponentFactory` and `SpringAwareUtBotComponentFactory` need to be carefully chosen.

```kotlin
interface UtBotComponentFactory {
    val utExecutionInstrumentation : Instrumentation<UtConcreteExecutionResult>

    fun createJavaValueProvider(idGenerator: IdentityPreservingIdGenerator<Int>, methodUnderTest: MethodId) : JavaValueProvider

    fun createCodeGenerator(...): CodeGenerator
    
    ...
}

class SimpleUtBotComponentFactory : UtBotComponentFactory {
    override val utExecutionInstrumentation = UtExecutionInstrumentation

    override fun createJavaValueProvider(idGenerator: IdentityPreservingIdGenerator<Int>, methodUnderTest: MethodId) = ValueProvider.of(
        defaultValueProviders(idGenerator)
    )

    override fun createCodeGenerator(...) = SimpleGenerator(...)
    
    ...
}

class SpringAwareUtBotComponentFactory(
    private val delegateFactory: UtBotComponentFactory,
    private val instantiationSettings: InstantiationSettings,
    private val beanDefinitions: List<BeanDefinitionData>
) : UtBotComponentFactory {
    override val utExecutionInstrumentation = SpringUtExecutionInstrumentation(
        delegateFactory.utExecutionInstrumentation,
        instantiationSettings,
        beanDefinitions
    )

    override fun createJavaValueProvider(idGenerator: IdentityPreservingIdGenerator<Int>, methodUnderTest: MethodId) = 
        delegateFactory.createJavaValueProvider(idGenerator).let {
                provider ->
            val relevantRepositories = concreteExecutor.getRelevantSpringRepositories(methodUnderTest.classId)
            logger.info { "Detected relevant repositories for class ${methodUnderTest.classId}: $relevantRepositories" }

            val generatedValueAnnotationClasses = SpringModelUtils.generatedValueClassIds.mapNotNull {
                @Suppress("UNCHECKED_CAST") // type system fails to understand that GeneratedValue is indeed an annotation
                utContext.classLoader.tryLoadClass(it.name) as Class<out Annotation>?
            }

            val generatedValueFieldIds =
                relevantRepositories
                    .map { it.entityClassId.jClass }
                    .flatMap { entityClass -> generateSequence(entityClass) { it.superclass } }
                    .flatMap { it.declaredFields.toList() }
                    .filter { field -> generatedValueAnnotationClasses.any { field.isAnnotationPresent(it) } }
                    .map { it.fieldId }
            logger.info { "Detected @GeneratedValue fields: $generatedValueFieldIds" }

            // spring should try to generate bean values, but if it fails, then object value provider is used for it
            val springBeanValueProvider = SpringBeanValueProvider(
                idGenerator,
                beanNameProvider = { classId ->
                    beanDefinitions
                        .filter { it.beanTypeFqn == classId.name }
                        .map { it.beanName }
                },
                relevantRepositories = relevantRepositories
            ).withFallback(ObjectValueProvider(idGenerator))
            provider
                .except { p -> p is ObjectValueProvider }
                .with(springBeanValueProvider)
                .with(ValueProvider.of(relevantRepositories.map { SavedEntityValueProvider(defaultIdGenerator, idGenerator) }))
                .with(ValueProvider.of(generatedValueFieldIds.map { FieldValueProvider(defaultIdGenerator, idGenerator) }))
        }

    override fun createCodeGenerator(...): CodeGenerator = SpringCodeGenerator(delegateFactory.createCodeGenerator(...), ...)
    
    ...
}
```

### PR DESCRIPTION
**Pull request by IlyaMuravjov**
## Description

Fixes #2358, #2226 

## How to test

Regression tests (this pull request should not change UtBot semantics at all)

## Self-check list

- [x] I've set the proper **labels** for my PR (at least, for category and component).
- [x] PR **title** and **description** are clear and intelligible.
- [x] I've added enough **comments** to my code, particularly in hard-to-understand areas.
- [x] The functionality I've repaired, changed or added is covered with **automated tests**.
- [ ] **Manual tests** have been provided optionally.
- [x] The **documentation** for the functionality I've been working on is up-to-date.

### CODE REVIEW

**Review by Damtev — APPROVED** [review]
LGTM

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/execution/UtExecutionInstrumentation.kt [hunk 1]
```diff
@@ -47,20 +47,39 @@ data class UtConcreteExecutionResult(
...
+        override fun create(): TInstrumentation = create(SimpleInstrumentationContext())
 
-    override fun init(pathsToUserClasses: Set<String>) {
-        UtExecutionInstrumentation.pathsToUserClasses.clear()
-        UtExecutionInstrumentation.pathsToUserClasses += pathsToUserClasses
+        fun create(instrumentationContext: InstrumentationContext): TInstrumentation
     }
+}
+
+class SimpleUtExecutionInstrumentation(
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/execution/UtExecutionInstrumentation.kt:75]: Let's extract it to a separate file

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/InvokeInstrumentation.kt [hunk 2]
```diff
@@ -117,4 +117,8 @@ class InvokeInstrumentation : Instrumentation<Result<*>> {
         protectionDomain: ProtectionDomain,
         classfileBuffer: ByteArray
     ) = null
+
+    class Factory : Instrumentation.Factory<Result<*>, InvokeInstrumentation> {
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/InvokeInstrumentation.kt:121]: Why not object?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/et/ExecutionTraceInstrumentation.kt [hunk 3]
```diff
@@ -57,4 +57,8 @@ class ExecutionTraceInstrumentation : Instrumentation<Trace> {
             classByteCode
         }
     }
+
+    class Factory : Instrumentation.Factory<Trace, ExecutionTraceInstrumentation> {
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/et/ExecutionTraceInstrumentation.kt:61]: Why not object?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/coverage/CoverageInstrumentation.kt [hunk 4]
```diff
@@ -94,6 +94,10 @@ object CoverageInstrumentation : Instrumentation<Result<*>> {
 
         return instrumenter.classByteCode
     }
+
+    class Factory : Instrumentation.Factory<Result<*>, CoverageInstrumentation> {
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/coverage/CoverageInstrumentation.kt:98]: Why not object?

File: utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/InvokeWithStaticsInstrumentation.kt [hunk 5]
```diff
@@ -89,6 +89,10 @@ class InvokeWithStaticsInstrumentation : Instrumentation<Result<*>> {
                 .forEach { it.withAccessibility { it.set(null, savedFields[it.name]) } }
         }
     }
+
+    class Factory : Instrumentation.Factory<Result<*>, InvokeWithStaticsInstrumentation> {
```
Comment (sergeypospelov) [inline utbot-instrumentation/src/main/kotlin/org/utbot/instrumentation/instrumentation/InvokeWithStaticsInstrumentation.kt:93]: Why not object?

**Review by sergeypospelov — CHANGES_REQUESTED** [review]

**Review by EgorkaKulikov — APPROVED** [review]

**Review by sergeypospelov — APPROVED** [review]

**Pull request merged**